File: cases/taint/php/mutual-recursive-transform-positive/mutual_recursive_transform_positive.php
<?php
function dfb_source(): int { // DFB-SOURCE: mutual-recursive-transform-input
    return 7;
}

function dfb_sink(int $value): void {} // DFB-SINK: mutual-recursive-transform-sink

function walkA(int $value, int $depth): int
{
    if ($depth === 0) {
        return $value;
    }
    $next = $value + 1;
    $recursive = walkB($next, $depth - 1); // DFB-WITNESS: mutual-recursive-transform-a-transfer
    return $recursive + 1; // DFB-WITNESS: mutual-recursive-transform-a-compose
}

function walkB(int $value, int $depth): int
{
    if ($depth === 0) {
        return $value; // DFB-WITNESS: mutual-recursive-transform-b-base
    }
    $next = $value + 1;
    $recursive = walkA($next, $depth - 1); // DFB-WITNESS: mutual-recursive-transform-b-transfer
    return $recursive + 1; // DFB-WITNESS: mutual-recursive-transform-b-compose
}

function run(): void
{
    dfb_sink(walkA(dfb_source(), 3));
}

File: cases/taint/scala/mutual-recursive-transform-positive/MutualRecursiveTransformPositive.scala
package dataflowbench

object MutualRecursiveTransformPositive {
  def dfb_source(): Int = { // DFB-SOURCE: mutual-recursive-transform-input
    7
  }

  def dfb_sink(value: Int): Unit = {} // DFB-SINK: mutual-recursive-transform-sink

  def walkA(value: Int, depth: Int): Int = {
    if (depth == 0) {
      value
    } else {
      val next = value + 1
      val recursive = walkB(next, depth - 1) // DFB-WITNESS: mutual-recursive-transform-a-transfer
      recursive + 1 // DFB-WITNESS: mutual-recursive-transform-a-compose
    }
  }

  def walkB(value: Int, depth: Int): Int = {
    if (depth == 0) {
      value // DFB-WITNESS: mutual-recursive-transform-b-base
    } else {
      val next = value + 1
      val recursive = walkA(next, depth - 1) // DFB-WITNESS: mutual-recursive-transform-b-transfer
      recursive + 1 // DFB-WITNESS: mutual-recursive-transform-b-compose
    }
  }

  def run(): Unit = {
    dfb_sink(walkA(dfb_source(), 3))
  }
}
